### ISSUE TITLE
Unreasonably low results, when using IAM dataset

### ISSUE THREAD

**Issue by mcmalzahar** [sushant097/Handwritten-Line-Text-Recognition-using-Deep-Learning-with-Tensorflow#13]
I'm also getting incredibly poor scores, character error rate of ~45%, word error rate of ~650%
Using the IAM lines dataset, all according to instructions.
Not using word beam search decoding or anything similar, just `python main.py --train`
Exact same results using either CPU or GPU (the only difference of course being that the GPU way is several times faster). The CTC loss scores and the error rate stop improving after the 7th or so epoch.
Tried increasing the file input size, but got considerably worse results.
Attached an image of an example word error rate while training.

![screenshot_1](https://user-images.githubusercontent.com/25721361/80967357-eaf04780-8e1e-11ea-9426-237bce64ca6f.png)

My specs are:
16 GB RAM, Ryzen 3700X CPU, NVIDIA RTX2070 SUPER GPU. Using Docker and TensorFlow 1.13.

_Originally posted by @mcmalzahar in https://github.com/sushant097/Handwritten-Line-Text-Recognition-using-Deep-Learning-with-Tensorflow/issues/4#issuecomment-623444144_

**Comment by rpro91** [comment]
I have also faced the same issue when I ran it last year.. you might look for some other model like https://github.com/lamhoangtung/LineHTR. I was able to run this one.

**Comment by mcmalzahar** [comment]
> I have also faced the same issue when I ran it last year.. you might look for some other model like https://github.com/lamhoangtung/LineHTR. I was able to run this one.

Unfortunately, that model has no instructions on how to use it on the IAM dataset, which is the main reason why I chose this one. I've generated a labels.json in the proper format to get IAM working, but still no luck...

**Comment by imsunnykr** [comment]
> I'm also getting incredibly poor scores, character error rate of ~45%, word error rate of ~650%
> Using the IAM lines dataset, all according to instructions.
> Not using word beam search decoding or anything similar, just `python main.py --train`
> Exact same results using either CPU or GPU (the only difference of course being that the GPU way is several times faster). The CTC loss scores and the error rate stop improving after the 7th or so epoch.
> Tried increasing the file input size, but got considerably worse results.
> Attached an image of an example word error rate while training.
> 
> ![screenshot_1](https://user-images.githubusercontent.com/25721361/80967357-eaf04780-8e1e-11ea-9426-237bce64ca6f.png)
> 
> My specs are:
> 16 GB RAM, Ryzen 3700X CPU, NVIDIA RTX2070 SUPER GPU. Using Docker and TensorFlow 1.13.
> 
> _Originally posted by @mcmalzahar in [#4 (comment)](https://github.com/sushant097/Handwritten-Line-Text-Recognition-using-Deep-Learning-with-Tensorflow/issues/4#issuecomment-623444144)_

Replace the batch Size to 50 from 10 under model.py. This will help in getting good recognition accuracy.

**Comment by imsunnykr** [comment]
> I'm also getting incredibly poor scores, character error rate of ~45%, word error rate of ~650%
> Using the IAM lines dataset, all according to instructions.
> Not using word beam search decoding or anything similar, just `python main.py --train`
> Exact same results using either CPU or GPU (the only difference of course being that the GPU way is several times faster). The CTC loss scores and the error rate stop improving after the 7th or so epoch.
> Tried increasing the file input size, but got considerably worse results.
> Attached an image of an example word error rate while training.
> 
> ![screenshot_1](https://user-images.githubusercontent.com/25721361/80967357-eaf04780-8e1e-11ea-9426-237bce64ca6f.png)
> 
> My specs are:
> 16 GB RAM, Ryzen 3700X CPU, NVIDIA RTX2070 SUPER GPU. Using Docker and TensorFlow 1.13.
> 
> _Originally posted by @mcmalzahar in [#4 (comment)](https://github.com/sushant097/Handwritten-Line-Text-Recognition-using-Deep-Learning-with-Tensorflow/issues/4#issuecomment-623444144)_

Change the batch size to 50 under model.py and train the model. This will help.

**Comment by vgangal101** [comment]
There is one aspect that doesn't make sense here. How is it possible to get a WER of above 650%? That clearly means the WER calculation is being done wrong. Is there gonna be a patch for this?

**Comment by sushant097** [comment]
@mcmalzahar I can't find the reason behind your issue. Please recheck the code what wrong is inside yours. I can't answer based on the results that you get. Please use Tensorflow 1.8.0.

**Comment by Tamoghna-Saha** [comment]
@vgangal101 don't know by this time whether you have resolved this issue of WER 650%.

I will suggest you use [this](https://github.com/githubharald/CTCWordBeamSearch/blob/master/py/Metrics.py) for metric evaluation.

Even in this code, there is an exception that hasn't been handled.

I observed that if the words are completely different, as in, ground truth is "hello" and predicted result is "by", then there is no overlap of common characters. In that situation, the number of incorrect characters exceeds than the ground truth characters. So the CER is also coming > 1 (or 100%) and if you look at the code for WER evaluation, we will be getting WER > 1.

So I did this.

```
    def getCER(self):
        "get character error rate"
        if self.edChars > self.numChars:
            return 1.0
        elif self.numChars > 0:
            return self.edChars / self.numChars
        else:
            return 1.0

    def getWER(self):
        "get word error rate"
        if self.edWords > self.numWords:
            return 1.0
        elif self.numWords > 0:
            return self.edWords / self.numWords
        else:
            return 1.0
```

**Comment by dinukakulatunga** [comment]
I'm getting this error when I try to run the main.py. please help me.
![image](https://user-images.githubusercontent.com/32608112/85404353-73b47580-b57c-11ea-99ab-293601e1da03.png)

**Comment by ab6995** [comment]
@mcmalzahar i faced the same issue you had in the first run.
Solution that worked for me:

- Increased the batch size to 128
- Some changes i made in pre process to do the alignments in file 

location -> src/DataLoader.py line(79-86)
```
 if len(lineSplit)== 9 :
    gtText_list = lineSplit[8].split('|')
 if len(lineSplit)>9:
    l = len(lineSplit)
    for i in range(8,l-1):
        lineSplit[8] = lineSplit[8] + "|" +lineSplit[i+1]
        i = i+1
    gtText_list = lineSplit[8].split('|')
```

**Comment by sushant097** [comment]
Inactive for long time.
